src/main/java/dev/jbang/net/KotlinManager.java

```diff
@@ -0,0 +1,83 @@
+package dev.jbang.net;
+
+import static dev.jbang.cli.BaseCommand.EXIT_UNEXPECTED_STATE;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import dev.jbang.Cache;
+import dev.jbang.Settings;
+import dev.jbang.cli.ExitException;
+import dev.jbang.util.JavaUtil;
+import dev.jbang.util.UnpackUtil;
+import dev.jbang.util.Util;
+
+public class KotlinManager {
+	private static final String KOTLIN_DOWNLOAD_URL = "https://github.com/JetBrains/kotlin/releases/download/v%s/kotlin-compiler-%s.zip";
+	public static final String DEFAULT_KOTLIN_VERSION = "1.5.0";
+
+	public static String resolveInKotlinHome(String cmd, String requestedVersion) {
+		Path kotlinHome = getKotlin(requestedVersion);
+		if (kotlinHome != null) {
+			if (Util.isWindows()) {
+				cmd = cmd + ".exe";
+			}
+			return kotlinHome.resolve("bin").resolve(cmd).toAbsolutePath().toString();
+		}
+		return cmd;
+	}
+
+	public static Path getKotlin(String requestedVersion) {
+		Path kotlinPath = getKotlinPath(requestedVersion);
+		if (!Files.isDirectory(kotlinPath)) {
+			kotlinPath = downloadAndInstallKotlin(requestedVersion);
+		}
+		return kotlinPath.resolve("kotlinc");
+	}
+
+	public static Path downloadAndInstallKotlin(String version) {
+		Util.infoMsg("Downloading Kotlin " + version + ". Be patient, this can take several minutes...");
+		String url = String.format(KOTLIN_DOWNLOAD_URL, version, version);
+		Util.verboseMsg("Downloading " + url);
+		Path kotlinDir = getKotlinPath(version);
+		Path kotlinTmpDir = kotlinDir.getParent().resolve(kotlinDir.getFileName().toString() + ".tmp");
+		Path kotlinOldDir = kotlinDir.getParent().resolve(kotlinDir.getFileName().toString() + ".old");
+		Util.deletePath(kotlinTmpDir, false);
+		Util.deletePath(kotlinOldDir, false);
+		try {
+			Path kotlinPkg = Util.downloadAndCacheFile(url);
+			Util.infoMsg("Installing Kotlin " + version + "...");
+			Util.verboseMsg("Unpacking to " + kotlinDir);
+			UnpackUtil.unpack(kotlinPkg, kotlinTmpDir);
+			if (Files.isDirectory(kotlinDir)) {
+				Files.move(kotlinDir, kotlinOldDir);
+			}
+			Files.move(kotlinTmpDir, kotlinDir);
+			Util.deletePath(kotlinOldDir, false);
+			return kotlinDir;
+		} catch (Exception e) {
+			Util.deletePath(kotlinTmpDir, true);
+			if (!Files.isDirectory(kotlinDir) && Files.isDirectory(kotlinOldDir)) {
+				try {
+					Files.move(kotlinOldDir, kotlinDir);
+				} catch (IOException ex) {
+					// Ignore
+				}
+			}
+			Util.errorMsg("Required Java version not possible to download or install. You can run with '--java "
+					+ JavaUtil.determineJavaVersion() + "' to force using the default installed Java.");
```

Comment:
this would need some other detection for kotlin

Reply:
I used to have a `determinKotlinVersion()` but it looks like it got refactored away.  I'm not sure this concept even makes sense for kotlin because it's not like we're running a KotlinVM here.  the version of kotlin is whatever is configured in the script or on the command line.  Other than that, there's nothing really left to discover, as such.

Reply:
there isn't even a explicit control of --kotlin yet thus for now remove this to avoid confusion. we can add back if/when usecase arise for the external control.